src/coreclr/vm/jitinterface.cpp

```diff
@@ -1721,6 +1721,54 @@ bool CEEInfo::isFieldStatic(CORINFO_FIELD_HANDLE fldHnd)
     return res;
 }
 
+int CEEInfo::getReadonlyStaticArrayLength(CORINFO_FIELD_HANDLE fldHnd)
```

Comment:
This model where we are always asking for properties of something stored in a field is not going to compose well.

I expect that one of the next things you will want to add is a way to inspect objectrefs wrapped in structs, say `static ImmutableArray<Something>`. Do we want to design the JIT/EE surface such that it is prepared for it?

It would need ability for the JIT/EE interface to return handles to objects. So asking for array length of an array stored in a field would be two step process: First step is to get the handle of the array, second step is to get its length.

Reply:
@jkotas am I correct that we'll have to somehow pin such handles for jit (since it works in the preemptive mode) and then rely on jit to unpin them? so essentially it will be 3 calls, e.g.:

```csharp
handle = vm->getPinnedObjectHandle(field);
int arrayLen = vm->getArrayLen(handle);
vm->Unpin(handle)
```
?

Reply:
I think it can be just:
```
handle = vm->getPinnedObjectHandle(field);
int arrayLen = vm->getArrayLen(handle);
```
The JIT/EE interface impl would keep track of all handles allocated during JITing a method and free up all of them once the JIT returns.

I do not think there would be any actual pinning. It would just allocate regular (non-pinning) GC handles, equivalent of
`var handle =(IntPtr)GCHandle.Create(...);`.

Reply:
One option is to skip allocating GC handles for frozen objects and use the lowest bit of the handle returned to the JIT to tell whether it is frozen object or regular object wrapped in handle. The code to wrap an object into a handle would look like this:

```
if (isFrozenObject(obj))
{
    handle = (CORINFO_OBJECT_HANDLE)obj;
}
else
{
   handle = (CORINFO_OBJECT_HANDLE)((TADDR)CreateHandle(obj) | 1);
}
```

It would avoid paying `CreateHandle` overhead for frozen objects.

Reply:
@jkotas thanks! will try. Last thing to clarify: I'll have to have a list of handles which I will need to release (all of them) once jit finishes, right? on the vm side.

Reply:
Yep